File: local-llama.py
# This spins up a local vllm server that is more production ready than the js one and is meant
# to be used in strong compute hardware when the entire model fits on VRAM
# It supports multiple concurrent requests, cancellation, and has a more robust protocol for error handling and responses. It also includes a simple secret key
# authentication mechanism to prevent unauthorized access.

# While this is more or less complete, it doesn't really have robust security measures in place for multiple users
# that are not trusted, so it should only be used with friends or a known group that has access to the "secret" file that contains
# the password for access the server

# this can also be used for production but it will need a proxy to handle proper security

import asyncio
import json
import secrets
import ssl
import sys
import time
import websockets
import base
from urllib.parse import urlparse, parse_qs
import os

PORT = 8765
HOST = '0.0.0.0'

# get the DEV=1 environment variable to make a simpler secret for development
DEV = os.getenv("DEV", "0") == "1"

END_TOKEN = None  # This will be set after loading the config

# ── Static info page ──────────────────────────────────────────────────────
# Load the HTML template once at startup. Visiting https://host:8765/ in a
# browser shows basic server status; this also gives users a way to manually
# accept the self-signed certificate so subsequent wss:// connections work.
_INDEX_HTML_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "index.html")
try:
    with open(_INDEX_HTML_PATH, "r", encoding="utf-8") as _f:
        INDEX_HTML_TEMPLATE = _f.read()
except Exception as _e:
    print(f"Warning: failed to load index.html template: {_e}")
    INDEX_HTML_TEMPLATE = "<html><body><h1>DreamServer</h1><p>(template missing)</p></body></html>"

SERVER_START_TIME = time.time()
ARG_CONFIG_PATH = ""  # set in __main__


def _html_escape(s: str) -> str:
    return (
        str(s)
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("'", "&#39;")
    )


def _format_uptime(seconds: float) -> str:
    s = int(seconds)
    d, s = divmod(s, 86400)
    h, s = divmod(s, 3600)
    m, sec = divmod(s, 60)
    parts = []
    if d:
        parts.append(f"{d}d")
    if h or d:
        parts.append(f"{h}h")
    if m or h or d:
        parts.append(f"{m}m")
    parts.append(f"{sec}s")
    return " ".join(parts)


def _render_index_html() -> str:
    replacements = {
        "PROTOCOL": "wss",
        "DEV_MODE": "DEV (insecure secret)" if DEV else "production",
        "SSL_MODE": "enabled",
        "MODEL_LOADED": "yes" if getattr(base, "MODEL", None) is not None else "no",
        "MODEL_PATH": _html_escape(getattr(base, "MODEL_PATH", "") or "(none)"),
        "CONFIG_PATH": _html_escape(ARG_CONFIG_PATH or "(none)"),
        "CONFIG_MODE": _html_escape(os.environ.get("CONFIG_MODE", "(see config file)")),
        "END_TOKEN": _html_escape(END_TOKEN or ""),
        "CONTEXT_WINDOW": str(getattr(base, "CONTEXT_WINDOW_SIZE", "?")),
        "GPU": _html_escape(os.environ.get("GPU", "auto")),
        "UPTIME": _format_uptime(time.time() - SERVER_START_TIME),
        "PROGRAM": "Python vLLM Server",
    }
    out = INDEX_HTML_TEMPLATE
    for key, value in replacements.items():
        out = out.replace("{{" + key + "}}", str(value))
    return out


async def handle_client(websocket):
    print('Client connected')

    await websocket.send(json.dumps({
        "type": "ready",
        "message": "Model is ready",
        "context_window_size": base.CONTEXT_WINDOW_SIZE,
        "supports_parallel_requests": True,
        "end_token": END_TOKEN
    }))

    loop = asyncio.get_event_loop()
    executor = None  # Use default executor (main thread)
    try:
        requestid_to_rid = {}  # Map internal request IDs to client rids for cancellation
        async for message in websocket:
            try:
                data = json.loads(message)
                rid = data.get('rid', 'no-rid')  # Use provided rid or default to 'no-rid'
                internal_request_id = None  # To track the internal request ID for this action, if applicable

                if (base.DEBUG):
                    print(f"Received message from client: {data}")

                payload = data['payload'] if 'payload' in data else None

                # Handle different actions
                if data.get('action') == 'infer':
                    if not payload:
                        raise ValueError("Invalid payload for infer")
                    
                    async for result in base.generate_completion(payload):
                        if 'error' in result:
                            loop.call_soon_threadsafe(asyncio.create_task, websocket.send(json.dumps({"type": "error", "rid": rid, "message": str(result['error'])})))
                            if internal_request_id is not None:
                                requestid_to_rid.pop(internal_request_id, None)  # Remove from active requests on error
                        elif 'token' in result:
                            loop.call_soon_threadsafe(asyncio.create_task, websocket.send(json.dumps({"type": "token", "rid": rid, "text": result['token']})))
                        elif 'done' in result and result['done']:
                            loop.call_soon_threadsafe(asyncio.create_task, websocket.send(json.dumps({"type": "done", "rid": rid})))
                            if internal_request_id is not None:
                                requestid_to_rid.pop(internal_request_id, None)  # Remove from active requests on done
                        elif 'request_id' in result:
                            internal_request_id = result['request_id']
                            requestid_to_rid[internal_request_id] = rid  # Track active request ID
                elif data.get('action') == 'analyze-prepare':
                    if not payload:
                        raise ValueError("Invalid payload for analyze-prepare")

                    async for result in base.prepare_analysis(payload):
                        if 'error' in result:
                            loop.call_soon_threadsafe(asyncio.create_task, websocket.send(json.dumps({"type": "error", "rid": rid, "message": str(result['error'])})))
                            if internal_request_id is not None:
                                requestid_to_rid.pop(internal_request_id, None)  # Remove from active requests on error
                        elif 'done' in result and result['done']:
                            loop.call_soon_threadsafe(asyncio.create_task, websocket.send(json.dumps({"type": "analyze-ready", "rid": rid})))
                            if internal_request_id is not None:
                                requestid_to_rid.pop(internal_request_id, None)  # Remove from active requests on done
                    
                elif data.get('action') == 'analyze-question':
                    if not payload:
                        raise ValueError("Invalid payload for analyze-question")
                    
                    async for result in base.run_question(payload):
                        if 'error' in result:
                            loop.call_soon_threadsafe(asyncio.create_task, websocket.send(json.dumps({"type": "error", "rid": rid, "message": str(result['error'])})))
                            if internal_request_id is not None:
                                requestid_to_rid.pop(internal_request_id, None)  # Remove from active requests on error
                        elif 'answer' in result:
                            loop.call_soon_threadsafe(asyncio.create_task, websocket.send(json.dumps({"type": "answer", "rid": rid, "text": result['answer']})))
                            if result.get('done'):
                                if internal_request_id is not None:
                                    requestid_to_rid.pop(internal_request_id, None)  # Remove from active requests on done
                        elif 'request_id' in result:
                            internal_request_id = result['request_id']
                            requestid_to_rid[internal_request_id] = rid  # Track active request ID
                elif data.get('action') == 'count-tokens':
                    if not payload or not isinstance(payload.get('text'), str):
                        raise ValueError("Invalid payload for count-tokens")
                    text = payload['text']
                    tokens = base.MODEL.tokenizer.encode(text)
                    await websocket.send(json.dumps({"type": "count", "rid": rid, "n_tokens": len(tokens)}))
                elif data.get('action') == 'cancel':
                    if not data.get('rid'):
                        raise ValueError("Missing rid for cancel action")
                    # Cancel the request with the given rid
                    cancel_rid = data['rid']
                    if (cancel_rid == 'no-rid'):
                        await websocket.send(json.dumps({"type": "error", "rid": rid, "message": "Cannot cancel request without a valid rid"}))
                        continue

                    for key, value in requestid_to_rid.items():
                        if value == cancel_rid:
                            internal_request_id = key
                            break

                    if internal_request_id is not None:
                        base.MODEL.abort_request(internal_request_id)
                        requestid_to_rid.pop(internal_request_id, None)  # Remove from active requests
            except Exception as e:
                print(str(e))
                await websocket.send(json.dumps({"type": "error", "rid": rid, "message": str(e)}))
    except websockets.ConnectionClosedOK:
        print('Client disconnected normally')
        for internalrid in requestid_to_rid.keys():
            base.MODEL.abort_request(internalrid)  # Stop the requests if the client disconnects
    except websockets.ConnectionClosedError as e:
        print(f'Client disconnected abnormally: code={e.code} reason={e.reason}')
        for internalrid in requestid_to_rid.keys():
            base.MODEL.abort_request(internalrid)  # Stop the requests if the client disconnects

async def main():
    ssl_context = None
    try:
        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
        ssl_context.load_cert_chain(certfile="cert.pem", keyfile="key.pem")
        print("SSL context created with cert.pem and key.pem")
    except Exception as e:
        print(f"Failed to create SSL context: {e}")
        sys.exit(1)
        
    server = await websockets.serve(handle_client, HOST, PORT, process_request=process_request, ssl=ssl_context)
    await server.serve_forever()

async def process_request(connection, request):
    parsed = urlparse(request.path)

    # Serve a small public info page on root, no auth required. This lets
    # users hit https://host:8765/ in a browser to accept the self-signed
    # cert so subsequent wss:// connections from the web app work.
    if parsed.path in ("/", "/index.html"):
        body = _render_index_html().encode("utf-8")
        headers = [
            ("Content-Type", "text/html; charset=utf-8"),
            ("Cache-Control", "no-store"),
            ("Content-Length", str(len(body))),
        ]
        return (200, headers, body)

    query_params = parse_qs(parsed.query)
    secret = query_params.get('secret', [None])[0]

    if not DEV:
        # read the local ./secret file for the expected API key
        try:
            with open("./secret", "r") as f:
                expected_secret = f.read().strip()
        except Exception as e:
            # create a new random API secret and save it to the file
            import secrets
            expected_secret = secrets.token_hex(64)
            with open("./secret", "w") as f:
                f.write(expected_secret)
            print(f"Generated new Secret key and saved to ./secret: {expected_secret}")
    else:
        expected_secret = "dev-secret-12345678900abcdef"  # Simple secret for development mode

    if secret != expected_secret:
        print(f"Unauthorized connection attempt with invalid secret")
        return (401, [], b'Unauthorized')  # Return 401 Unauthorized if the secret key is incorrect

    # Returning None continues the WebSocket handshake
    return None

if __name__ == "__main__":
    argv = sys.argv[1:]
    if len(argv) < 1:
        print("Please provide a config path as the first argument.", file=sys.stderr)
        sys.exit(1)

    ARG_CONFIG_PATH = argv[0]

    if not DEV:
        try:
            with open("./secret", "r") as f:
                print("Using Secret key from ./secret for authentication:")
        except Exception as e:
            print("No existing secret key found. A new one will be generated and saved to ./secret.")
            
            expected_secret = secrets.token_hex(64)
            with open("./secret", "w") as f:
                f.write(expected_secret)
            print(f"Generated new Secret key and saved to ./secret: {expected_secret}")

    print("DEBUG mode:", base.DEBUG)
    print("DEV mode:", DEV)
    END_TOKEN = base.load_config(argv[0])["end_token"]  # Load the config and set the END_TOKEN based on the model mode
    asyncio.run(main())

File: local-llama.js
// this is an extremely simple websocket server that runs a llama.cpp instance locally
// that can only handle one request at a time

// this is very good for consumer grade hardware that can't run anything in parallell
// and the model runs locally in the same machine

import { WebSocketServer } from "ws";
import { CONTROLLER, MODEL, MODEL_PATH, generateCompletion, prepareAnalysis, runQuestion, loadConfig } from "./base.js";
import { readFileSync, writeFileSync, existsSync } from "fs";
import { randomBytes } from "crypto";
import { URL } from "url";
import { createServer as createHttpsServer } from "https";
import { createServer as createHttpServer } from "http";

const DEV = process.env.DEV === "1";
const SSL = process.env.SSL === "1";

const argv = process.argv.slice(2);
if (argv.length < 1) {
    console.error("Please provide a model path as the first argument.");
    process.exit(1);
}

const END_TOKEN = (await loadConfig(argv[0])).endToken;

let expectedSecret;
if (!DEV) {
    if (existsSync("./secret")) {
        expectedSecret = readFileSync("./secret", "utf-8").trim();
        console.log("Using Secret key from ./secret for authentication");
    } else {
        expectedSecret = randomBytes(64).toString("hex");
        writeFileSync("./secret", expectedSecret);
        console.log(`Generated new Secret key and saved to ./secret: ${expectedSecret}`);
    }
} else {
    expectedSecret = "dev-secret-12345678900abcdef";
}

console.log("DEV mode:", DEV);
console.log("SSL mode:", SSL);
console.log(`Starting Local LLaMA WebSocket Server, listening on ${SSL ? "wss" : "ws"}://0.0.0.0:8765`);

const verifyClient = (info) => {
    const url = new URL(info.req.url, `http://${info.req.headers.host}`);
    const secret = url.searchParams.get("secret");
    if (secret !== expectedSecret) {
        console.log("Unauthorized connection attempt with invalid secret");
        return false;
    }
    console.log("Client authenticated successfully");
    return true;
};

let server;
if (SSL) {
    try {
        server = createHttpsServer({
            cert: readFileSync("./cert.pem"),
            key: readFileSync("./key.pem"),
        });
    } catch (e) {
        console.error("Failed to create SSL server:", e.message);
        process.exit(1);
    }
} else {
    server = createHttpServer();
}

let CONTEXT_WINDOW_SIZE = 2048 * 4; // 8k context
if (process.env.CONTEXT_WINDOW_SIZE) {
    const envSize = parseInt(process.env.CONTEXT_WINDOW_SIZE);
    if (!isNaN(envSize) && envSize > 0) {
        CONTEXT_WINDOW_SIZE = envSize;
    }
}

// ── Static info page ──────────────────────────────────────────────────────
// Load the HTML template once at startup. Visiting http(s)://host:8765/ in a
// browser shows basic server status; this also gives users a way to manually
// accept the self-signed certificate so subsequent wss:// connections work.
import { fileURLToPath } from "url";
import { dirname, join } from "path";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const INDEX_HTML_TEMPLATE = readFileSync(join(__dirname, "index.html"), "utf-8");
const SERVER_START_TIME = Date.now();
const ARG_CONFIG_PATH = argv[0];

/**
 * @param {Record<string, string>} replacements
 */
function renderIndexHtml(replacements) {
    return INDEX_HTML_TEMPLATE.replace(/\{\{(\w+)\}\}/g, (_, key) =>
        Object.prototype.hasOwnProperty.call(replacements, key)
            ? String(replacements[key])
            : `{{${key}}}`,
    );
}

/**
 * @param {number} ms
 */
function formatUptime(ms) {
    const s = Math.floor(ms / 1000);
    const d = Math.floor(s / 86400);
    const h = Math.floor((s % 86400) / 3600);
    const m = Math.floor((s % 3600) / 60);
    const sec = s % 60;
    const parts = [];
    if (d) parts.push(`${d}d`);
    if (h || d) parts.push(`${h}h`);
    if (m || h || d) parts.push(`${m}m`);
    parts.push(`${sec}s`);
    return parts.join(" ");
}

function escapeHtml(str) {
    return String(str)
        .replace(/&/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;")
        .replace(/"/g, "&quot;")
        .replace(/'/g, "&#39;");
}

server.on("request", (req, res) => {
    // Only the root path serves the info page; everything else is 404.
    const url = new URL(req.url || "/", `http://${req.headers.host}`);
    if (url.pathname !== "/" && url.pathname !== "/index.html") {
        res.statusCode = 404;
        res.setHeader("Content-Type", "text/plain; charset=utf-8");
        res.end("Not found");
        return;
    }

    // MODEL_PATH and MODEL are live ES module bindings, so we always read
    // their current values here without needing to re-import.
    const html = renderIndexHtml({
        PROTOCOL: SSL ? "wss" : "ws",
        DEV_MODE: DEV ? "DEV (insecure secret)" : "production",
        SSL_MODE: SSL ? "enabled" : "disabled",
        MODEL_LOADED: MODEL ? "yes" : "no",
        MODEL_PATH: escapeHtml(MODEL_PATH || "(none)"),
        CONFIG_PATH: escapeHtml(ARG_CONFIG_PATH || "(none)"),
        CONFIG_MODE: escapeHtml(process.env.CONFIG_MODE || "(see config file)"),
        END_TOKEN: escapeHtml(END_TOKEN || ""),
        CONTEXT_WINDOW: String(CONTEXT_WINDOW_SIZE),
        GPU: escapeHtml(process.env.GPU || "auto"),
        UPTIME: formatUptime(Date.now() - SERVER_START_TIME),
        PROGRAM: "Node.js Local LLaMA Server",
    });
    res.statusCode = 200;
    res.setHeader("Content-Type", "text/html; charset=utf-8");
    res.setHeader("Cache-Control", "no-store");
    res.end(html);
});

const wss = new WebSocketServer({ server, verifyClient });
server.listen(8765, '0.0.0.0');

/**
 * @type {Promise<void>}
 */
let lastGenerationPromise = Promise.resolve();

wss.on('connection', (ws) => {
    console.log('Client connected');

    ws.send(JSON.stringify({
        type: 'ready',
        message: 'Model is ready',
        context_window_size: CONTEXT_WINDOW_SIZE,
        supports_parallel_requests: false,
        end_token: END_TOKEN,
    }));

    ws.on('message', async (message) => {
        let rid = "no-rid";
        try {
            // @ts-ignore
            const data = JSON.parse(message);
            rid = data.rid || "no-rid";

            // Handle different actions
            if (data.action === 'infer') {
                if (!data.payload) {
                    throw new Error("Invalid payload for infer");
                }
                lastGenerationPromise = lastGenerationPromise.catch(() => {}).then(() => {
                    return generateCompletion(data.payload, (text) => {
                        ws.send(JSON.stringify({ type: 'token', rid, text }));
                    }, () => {
                        ws.send(JSON.stringify({ type: 'done', rid }));
                    }, (error) => {
                        ws.send(JSON.stringify({ type: 'error', rid, message: error.message }));
                    });
                });
                await lastGenerationPromise;
            } else if (data.action === 'analyze-prepare') {
                if (!data.payload) {
                    throw new Error("Invalid payload for analyze-prepare");
                }
                lastGenerationPromise = lastGenerationPromise.catch(() => {}).then(() => {
                    return prepareAnalysis(data.payload, () => {
                        ws.send(JSON.stringify({ type: 'analyze-ready', rid }));
                    }, (error) => {
                        ws.send(JSON.stringify({ type: 'error', rid, message: error.message }));
                    });
                });
                await lastGenerationPromise;
            } else if (data.action === 'analyze-question') {
                if (!data.payload) {
                    throw new Error("Invalid payload for analyze-question");
                }
                lastGenerationPromise = lastGenerationPromise.catch(() => {}).then(() => {
                    return runQuestion(data.payload, (text) => {
                        ws.send(JSON.stringify({ type: 'answer', rid, text }));
                    }, (error) => {
                        ws.send(JSON.stringify({ type: 'error', rid, message: error.message }));
                    });
                });
                await lastGenerationPromise;
            } else if (data.action === 'count-tokens') {
                if (!data.payload || typeof data.payload.text !== "string") {
                    throw new Error("Invalid payload for count-tokens");
                }
                const text = data.payload.text;
                const tokens = MODEL.tokenize(text);
                ws.send(JSON.stringify({ type: 'count', rid, n_tokens: tokens.length }));
            } else if (data.action === 'cancel') {
                if (!data.rid) {
                    throw new Error("Missing rid for cancel action");
                }
                if (data.rid === "no-rid") {
                    throw new Error("Cannot cancel request with no rid");
                }
                if (CONTROLLER) {
                    CONTROLLER.abort();
                    // Cancel the request with the given rid
                    // This will depend on how you track active requests and their IDs
                    ws.send(JSON.stringify({ type: "cancelled", rid: data.rid }));
                } else {
                    ws.send(JSON.stringify({ type: "error", rid, message: "No active controller to cancel" }));
                }
            } else {
                throw new Error("Unknown action: " + data.action);
            }
        } catch (e) {
            // @ts-ignore
            console.log(e.message);
            // @ts-ignore
            ws.send(JSON.stringify({ type: 'error', rid, message: e.message }));
        }
    });

    ws.on('close', () => {
        console.log('Client disconnected');
        if (CONTROLLER) {
            CONTROLLER.abort();
        }
    });
});